core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -900,13 +902,12 @@ private void validateStrongPolicy(RexNode rexNode) {
     case ANY:
       List<RexNode> operands = ((RexCall) rexNode).getOperands();
       if (rexNode.getType().isNullable()) {
-        assert operands.stream()
-            .map(RexNode::getType)
-            .anyMatch(RelDataType::isNullable);
+        // Ignores the nullability change when all the operands are literals
+        // which come from the simplification.
+        assert operands.stream().map(RexNode::getType).anyMatch(RelDataType::isNullable)
+            || operands.stream().allMatch(p -> RexUtil.isLiteral(p, false));
```

Comment:
This looks strange, and it contradicts `Policy.ALL` definition

> This kind of expression is null if and only if at least one of its arguments is null

What do literals do here?

Reply:
During the simplification, we drop the useless case `cast(1 as nullable INT)` which breaks the nullability. We have 2 choices:
1. Do not make any simplification that breaks the nullability
2. Make the validation strategy more relaxed

I choose the latter.

Reply:
See https://issues.apache.org/jira/browse/CALCITE-4387
Simplify should be allowed to **both** widen and narrowing nullability.

I believe we must not alter the definition of `nullness policy`here.

Reply:
The `nullness policy` is always there, we should make the validation adapter to the simplification.

Reply:
`case ANY` is defined as `return type is nullable if and only if one of the arguments is null`.
It allows no exceptions like "ok, the return type might be null if one of the arguments is literal".

That is why I believe `RexUtil.isLiteral` must not be here.

Reply:
I agree with @vlsi. We should not change the type inference. In fact, type inference is working perfectly.

I think we should drop the requirement that `RexSimplify` does not change nullability. I know this conflicts with the requirement that `RelOptRule` does not change row type, which is a requirement we want to keep. The solution may be to add a cast (or some other operator) to say 'treat this not-null expression as nullable'.

Reply:
> I know this conflicts with the requirement that RelOptRule does not change row type,

There's no conflict.
The rules that need to match nullability should use `RexSimplify#simplifyPreservingType(RexNode e, RexUnknownAs unknownAs, boolean matchNullability = true)`

In other words, the rules should use `simplifyPreservingType`, and regular `simplify` calls should be free to alter nullability.

Reply:
Yeah, seems we have made conclusion that:
1. We should keep the routine that `RelOptRule` does not change nullability
2. The `RexSimplify` can change the `RexNode` nullability.

Reply:
3. Assertion at line 907 must be kept intact.

Reply:
Then we should never use `ReduceExpressionsRule.Config.matchNullability`, it should always return true.

I can not figure out a way to fix if we keep the assertion at line 907 as it is.

Reply:
`rexBuilder.makeCall(e.getType(), e.getOperator(), operands);` above must be updated to infer the proper type instead of `e.getType()`.

PS. I have already done all of that in https://github.com/apache/calcite/pull/2250

Reply:
> rexBuilder.makeCall(e.getType(), e.getOperator(), operands); above must be updated to infer the proper type instead of e.getType().

The problem is that it is hard to keep node type equivalence during simplification anymore, and the `ReduceExpressionsRule.Config.matchNullability` becomes meaningless.

Reply:
`ReduceExpressionsRule.Config.matchNullability` refers to the nullability of the overall expression rather than the nullability of the internal nodes.